File: ChatSupport/src/main/java/com/vantagecircle/chatapp/activity/SignupActivity.java
package com.vantagecircle.chatapp.activity;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatEditText;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.google.firebase.iid.FirebaseInstanceId;
import com.vantagecircle.chatapp.R;
import com.vantagecircle.chatapp.services.SupportService;
import com.vantagecircle.chatapp.core.AuthHandler;
import com.vantagecircle.chatapp.core.SetDataHandler;
import com.vantagecircle.chatapp.core.interfaceC.ResultInterface;
import com.vantagecircle.chatapp.model.RoomM;
import com.vantagecircle.chatapp.utils.Constants;
import com.vantagecircle.chatapp.model.UserM;
import com.vantagecircle.chatapp.pref.SharedPrefM;
import com.vantagecircle.chatapp.utils.ToolsUtils;

import java.util.ArrayList;

/**
 * Created by bapidas on 10/07/17.
 */

public class SignupActivity extends AppCompatActivity {
    private static final String TAG = LoginActivity.class.getSimpleName();
    private Activity activity;
    private Context mContext;
    private EditText usernameEdit, passwordEdit, fullNameEdit;
    private Button btnSignup;
    private String username, password, fullName;
    private ProgressDialog progressDialog;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activity = this;
        mContext = getApplicationContext();
        setContentView(R.layout.activity_signup);
        initView();
        initListener();
    }

    private void initView() {
        usernameEdit = (AppCompatEditText) findViewById(R.id.usernameEdit);
        passwordEdit = (AppCompatEditText) findViewById(R.id.passwordEdit);
        fullNameEdit = (AppCompatEditText) findViewById(R.id.fullNameEdit);
        btnSignup = (Button) findViewById(R.id.btnSignup);
    }

    private void initListener() {
        btnSignup.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (usernameEdit.getText().toString().length() == 0 &&
                        passwordEdit.getText().toString().length() == 0 &&
                        fullNameEdit.getText().toString().length() == 0) {
                    Toast.makeText(mContext, "All Details is cannot be blank",
                            Toast.LENGTH_SHORT).show();
                } else if (usernameEdit.getText().toString().length() == 0) {
                    Toast.makeText(mContext, "Username cannot be blank",
                            Toast.LENGTH_SHORT).show();
                } else if (passwordEdit.getText().toString().length() == 0) {
                    Toast.makeText(mContext, "Password cannot be blank",
                            Toast.LENGTH_SHORT).show();
                } else if (fullNameEdit.getText().toString().length() == 0) {
                    Toast.makeText(mContext, "Full name cannot be blank",
                            Toast.LENGTH_SHORT).show();
                } else {
                    if (ToolsUtils.isNetworkAvailable(mContext)) {
                        ToolsUtils.hideKeyboard(activity);
                        progressDialog = new ProgressDialog(activity);
                        progressDialog.setMessage("Validating fields and data");
                        progressDialog.show();
                        userSignup();
                    } else {
                        Toast.makeText(mContext, "Please check your internet connection",
                                Toast.LENGTH_SHORT).show();
                    }
                }
            }
        });
    }

    private void userSignup() {
        username = usernameEdit.getText().toString();
        password = passwordEdit.getText().toString();
        fullName = fullNameEdit.getText().toString();

        AuthHandler authHandler = new AuthHandler();
        authHandler.setFirebaseAuth(SupportService.getAuthInstance());
        authHandler.performSignup(username, password, new ResultInterface() {
            @Override
            public void onSuccess(String t) {
                progressDialog.setMessage("Creating account");
                setupData();
            }

            @Override
            public void onFail(String e) {
                if (progressDialog != null && progressDialog.isShowing()) {
                    progressDialog.dismiss();
                }
                Toast.makeText(mContext, e, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void setupData() {
        final UserM userM = new UserM();
        userM.setUserId(SupportService.getUserInstance().getUid());
        userM.setUsername(username);
        userM.setFullName(fullName);
        String fcmToken;
        if (new SharedPrefM(mContext).getString(Constants.FIREBASE_TOKEN) != null) {
            fcmToken = new SharedPrefM(mContext).getString(Constants.FIREBASE_TOKEN);
        } else {
            fcmToken = FirebaseInstanceId.getInstance().getToken();
            new SharedPrefM(SupportService.getInstance()).saveString(Constants.FIREBASE_TOKEN, fcmToken);
        }
        userM.setFcmToken(fcmToken);
        userM.setLastSeenTime(System.currentTimeMillis());
        userM.setNotificationCount(0);
        userM.setOnline(true);
        userM.setUserType(Constants._USER);
        userM.setRoomMArrayList(getStaticRooms());

        SetDataHandler setDataHandler = new SetDataHandler();
        setDataHandler.setDatabaseReference(SupportService.getUserReference().child(userM.getUserId()));
        setDataHandler.insertData(userM, new ResultInterface() {
            @Override
            public void onSuccess(String t) {
                SupportService.init(getApplicationContext(), userM);
                if (progressDialog != null && progressDialog.isShowing()) {
                    progressDialog.dismiss();
                }
                Toast.makeText(mContext, "Account is created successfully",
                        Toast.LENGTH_SHORT).show();
                Intent intent = new Intent(SignupActivity.this, UserActivity.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                startActivity(intent);
                finish();
            }

            @Override
            public void onFail(String e) {
                SupportService.getUserInstance().delete();
                if (progressDialog != null && progressDialog.isShowing()) {
                    progressDialog.dismiss();
                }
                Toast.makeText(mContext, e, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private ArrayList<RoomM> getStaticRooms(){
        ArrayList<RoomM> arrayList = new ArrayList<>();
        RoomM roomM = new RoomM();
        roomM.setRoomId("CAOL5K");
        roomM.setRoomName("CAOL5K_QWERTY");
        arrayList.add(roomM);
        return arrayList;
    }

    @Override
    protected void onResume() {
        Log.d(TAG, "onResume");
        super.onResume();
    }

    @Override
    protected void onPause() {
        Log.d(TAG, "onPause");
        super.onPause();
    }

    @Override
    protected void onStart() {
        Log.d(TAG, "onStart");
        super.onStart();
    }

    @Override
    protected void onStop() {
        Log.d(TAG, "onStop");
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        Log.d(TAG, "onDestroy");
        super.onDestroy();
    }
}

File: app/src/main/java/com/bapidas/chattingapp/ui/main/SignupActivity.kt
package com.bapidas.chattingapp.ui.main

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatEditText
import com.bapidas.chattingapp.ChatApplication
import com.bapidas.chattingapp.R
import com.bapidas.chattingapp.data.core.AuthHandler
import com.bapidas.chattingapp.data.core.SetDataHandler
import com.bapidas.chattingapp.data.core.callbacks.ResultInterface
import com.bapidas.chattingapp.data.model.ChatRoom
import com.bapidas.chattingapp.data.model.User
import com.bapidas.chattingapp.data.pref.SharedPrefHelper
import com.bapidas.chattingapp.utils.Constants
import com.bapidas.chattingapp.utils.ToolsUtils.hideKeyboard
import com.bapidas.chattingapp.utils.ToolsUtils.isNetworkAvailable
import com.google.firebase.iid.FirebaseInstanceId
import java.util.*

/**
 * Created by bapidas on 10/07/17.
 */
class SignupActivity : AppCompatActivity() {
    lateinit var usernameEdit: EditText
    lateinit var passwordEdit: EditText
    lateinit var fullNameEdit: EditText
    lateinit var btnSignup: Button

    private var username: String = ""
    private var password: String = ""
    private var fullName: String = ""

    private var progressDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_signup)
        initView()
        initListener()
    }

    private fun initView() {
        usernameEdit = findViewById<View>(R.id.usernameEdit) as AppCompatEditText
        passwordEdit = findViewById<View>(R.id.passwordEdit) as AppCompatEditText
        fullNameEdit = findViewById<View>(R.id.fullNameEdit) as AppCompatEditText
        btnSignup = findViewById<View>(R.id.btnSignup) as Button
    }

    private fun initListener() {
        btnSignup.setOnClickListener {
            if (usernameEdit.text.toString().isEmpty() && passwordEdit.text.toString().isEmpty()
                    && fullNameEdit.text.toString().isEmpty()) {
                Toast.makeText(this, "All Details is cannot be blank",
                        Toast.LENGTH_SHORT).show()
            } else if (usernameEdit.text.toString().isEmpty()) {
                Toast.makeText(this, "Username cannot be blank",
                        Toast.LENGTH_SHORT).show()
            } else if (passwordEdit.text.toString().isEmpty()) {
                Toast.makeText(this, "Password cannot be blank",
                        Toast.LENGTH_SHORT).show()
            } else if (fullNameEdit.text.toString().isEmpty()) {
                Toast.makeText(this, "Full name cannot be blank",
                        Toast.LENGTH_SHORT).show()
            } else {
                if (isNetworkAvailable(this)) {
                    hideKeyboard(this)
                    progressDialog = ProgressDialog(this)
                    progressDialog?.setMessage("Validating fields and data")
                    progressDialog?.show()
                    userSignup()
                } else {
                    Toast.makeText(this, "Please check your internet connection",
                            Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun userSignup() {
        username = usernameEdit.text.toString()
        password = passwordEdit.text.toString()
        fullName = fullNameEdit.text.toString()
        val authHandler = AuthHandler()
        authHandler.firebaseAuth = ChatApplication.applicationContext().authInstance
        authHandler.performSignup(username, password, object : ResultInterface {
            override fun onSuccess(t: String) {
                progressDialog?.setMessage("Creating account")
                setupData()
            }

            override fun onFail(e: String) {
                if (progressDialog != null && progressDialog?.isShowing == true) {
                    progressDialog?.dismiss()
                }
                Toast.makeText(this@SignupActivity, e, Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun setupData() {
        val userId = ChatApplication.applicationContext().userInstance?.uid.orEmpty()
        val username = username
        val fullName = fullName
        val fcmToken: String
        if (SharedPrefHelper(this).getString(Constants.FIREBASE_TOKEN) != null) {
            fcmToken = SharedPrefHelper(this).getString(Constants.FIREBASE_TOKEN).orEmpty()
        } else {
            fcmToken = FirebaseInstanceId.getInstance().token.orEmpty()
            SharedPrefHelper(this).saveString(Constants.FIREBASE_TOKEN, fcmToken)
        }
        val lastSeenTime = System.currentTimeMillis()
        val isOnline = true
        val userType = Constants.USER
        val roomMArrayList = staticChatRooms

        val userM = User(userId, username, fullName, fcmToken,
                lastSeenTime = lastSeenTime,
                isOnline = isOnline,
                userType = userType,
                chatRoomArrayList = roomMArrayList
        )
        val setDataHandler = SetDataHandler()
        setDataHandler.databaseReference = ChatApplication.applicationContext().userReference.child(userM.userId)
        setDataHandler.insertData(userM, object : ResultInterface {
            override fun onSuccess(t: String) {
                ChatApplication.applicationContext().user = userM
                if (progressDialog != null && progressDialog?.isShowing == true) {
                    progressDialog?.dismiss()
                }
                Toast.makeText(this@SignupActivity, "Account is created successfully",
                        Toast.LENGTH_SHORT).show()
                val intent = Intent(this@SignupActivity, UserActivity::class.java)
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                startActivity(intent)
                finish()
            }

            override fun onFail(e: String) {
                ChatApplication.applicationContext().userInstance?.delete()
                if (progressDialog != null && progressDialog?.isShowing == true) {
                    progressDialog?.dismiss()
                }
                Toast.makeText(this@SignupActivity, e, Toast.LENGTH_SHORT).show()
            }
        })
    }

    private val staticChatRooms: ArrayList<ChatRoom>
        get() {
            val arrayList = ArrayList<ChatRoom>()
            val roomId = "CAOL5K"
            val roomName = "CAOL5K_QWERTY"
            arrayList.add(ChatRoom(roomId, roomName))
            return arrayList
        }
}